### PR TITLE
Expose "get OCI manifest digest" callback

### PR DESCRIPTION
**Pull request by flavio**
Expose the get OCI manifest digest callback

Allow the Wasm guests to obtain the manifest's digest of
an OCI object (be it a container or an artifact).

Fetching information from a OCI registry requires the execution
of asynchronous code. Unfortunately the `host_callback` function
is a synchronous function.

Creating a dedicated tokio runtime (even by reusing the "current thread"
of the `host_callback` function) is not going to work. This would cause
tokio to cause a `panic` at runtime.

The solution of this problem consists of using a tokio channel that can
send the evaluation request from the synchronous world (the `host_callback`
function) to an `asynchronous` "worker" (something created with a tokio
`spawn` for example).

The response from the asynchronous world is then sent back to the
synchronous one by using a dedicated tokio `oneshot` channel.

## Changes

On top of exposing the new waPC host function, and adding all the tokio
channels. The following changes have been done:

  * The creation of a `PolicyEvaluator` is now done by using a dedicated
    builder: `PolicyEvaluatorBuilder`.
  * The code executing `PolicyEvaluator` (kwctl, policy-server) must
    prepare the worker that can handle the evaluation requests coming
    from the waPC guests.

### CODE REVIEW

**Comment by flavio** [comment]
This is now ready for review. We have to wait for the new version of `policy-fetcher` to be tagged, before we can merge this code

File: src/callback_requests.rs
Comment (ereslibre) [inline src/callback_requests.rs:18]:
```suggestion
        /// String pointing to the object (e.g.: `registry.testing.lan/busybox:1.0.0`)
```

File: src/callback_requests.rs
Comment (ereslibre) [inline src/callback_requests.rs:14]: What do you think about `CallbackRequest` being parameterized with the response type and having a two-level request, given every enum variant will have the response. Pseudorust:

```rust
pub enum CallbackRequestType {
  OciManifestDigest { image: String },
}

pub struct CallbackRequest<R> {
  request: CallbackRequestType,
  response_channel: oneshot::Sender<Result<R>>,
}
```

File: src/policy_evaluator.rs
Comment (ereslibre) [inline src/policy_evaluator.rs:115]: We can remove this comment given it's not part of public API

File: src/policy_evaluator_builder.rs
Comment (ereslibre) [inline src/policy_evaluator_builder.rs:10]: I must confess I love this pattern in Rust.

File: src/runtimes/wapc.rs
Comment (ereslibre) [inline src/runtimes/wapc.rs:56]: What is the expectation with burrego based policies? Do we plan to support something like this on those?

**Review by ereslibre — COMMENTED** [review]
Huge step in a great direction. Love it.

File: src/policy_evaluator.rs
Comment (flavio) [reply src/policy_evaluator.rs:115]: I left if on purpose for us, users of the internal `pub(crate)` method

File: src/runtimes/wapc.rs
Comment (flavio) [reply src/runtimes/wapc.rs:56]: I don't know, maybe in the future... but right now there's no way for them to reach out to the host. The only way would be to have us define some new Rego built-in and then plug that into the `CallbackHandler`

File: src/callback_requests.rs
Comment (flavio) [reply src/callback_requests.rs:14]: This is a really good idea. I'll change that

Comment (flavio) [reply src/callback_requests.rs]: Done

**Comment by flavio** [comment]
I think this could be the last round of reviews

**Review by viccuad — APPROVED** [review]
LGTM!

File: Cargo.toml
Comment (ereslibre) [inline Cargo.toml:28]: I am only a bit concerned about policy-evaluator depending on policy-fetcher. This is not a blocker, but I would prefer that they don't depend on each other, maybe by splitting common behaviors in other crate that both depend on.

**Review by ereslibre — APPROVED** [review]
LGTM, just a comment regarding dependencies, but I think we can tackle at a later time.

File: Cargo.toml
Comment (flavio) [reply Cargo.toml:28]: The main issue we have is: we need to interact with a container registry, to do that we need a docker configuration file and a file defining the sources (what is insecure, what is using a self-signed certificate,...).

These data types (`DockerConfig`, `Sources`) are defined inside of policy-fetcher. This crate has also all the conversion functions that can turn these data into something that `oci-distribution` can understand.

I have to admit my head :boom: every time I have to do all these conversions... And I had to do basically the same thing with the sigstore-rs crate, in order to not expose the mutable (as in they could change over the time) `oci-distribution` types into my public API. It's just a gigantic mess...

Sorry about the rant, I really don't have a better idea about how to tackle that

Comment (ereslibre) [reply Cargo.toml:28]: It's fine, in this case I prefer to have it working, and see if we can refactor later.

**Comment by flavio** [comment]
The last commit I pushed introduces the usage of a cache to temporarily store the results of the "get OCI manifest digest" invocations. 

I did that to address a significant performance drop inside of the PolicyServer.

# The testing scenario

I've ran our [load test](https://github.com/kubewarden/load-testing) against a local instance of PolicyServer. The PolicyServer was running uncontainerized and without TLS termination.

The PolicyServer was started with:
  * 2 worker threads to evaluate Wasm modules
  * 4 policies loaded, 1 of them used the "get OCI manifest digest" callback

The policy fetching an OCI manifest would reach a local docker registry running inside of my network.
The registry doesn't have authentication, but is TLS terminated using a certificate signed by trusted CA. 

## No policies using the OCI callback

To set a baseline, I configured the load tests to **not** reach the endpoint of the only policy using the OCI callback.

In this scenario the PolicyServer was able to process ~1500 requests per second.

## Hit the policy using the OCI callback - no caching

I updated the load test to ensure all the loaded policies, including the one using the OCI callback, could be hit.
It's important to stress that locust balances the request across all the different endpoints.

In this scenario the PolicyServer was able to process 15-20 requests per second (:exclamation: :exclamation: )

## Debugging the issue

@ereslibre and @viccuad raised some questions about the way the Callback Handler loop is managed inside of PolicyServer (see [this comment](https://github.com/kubewarden/policy-server/pull/126#discussion_r755281261)).

The CallbackHandler loop is not doing any blocking operation. It uses `await` and performs a`tokcio::select!` which is not causing the tokio runtime to block.

To confirm the performance issue is not caused by the CallbackHandler loop I did the following tests:
  * Performance are good when the CallbackHandler loop is running, but no request is sent to it. This is exactly the baseline scenario pictured before. The PolicyServer performance with the CallbackHandler loop running and without it (the code on the current `main` branch) are the same.
  * I replaced the code calculating the OCI digest to return a hard-coded string instead of reaching out to the actual registry. The load tests (hitting all the policies ofc) showed perfectly fine numbers

Thinking more about the issue, this is what actually happens:

* A POST request is received by the http server code of PolicyServer. This code runs inside of a tokio Runtime.
* The JSON request to be evaluated is sent from the API handler to the worker thread pool:
  * The communication between the async world (API handler) and the sync world (sync) happens over a tokio channel
  * The async world `await` on a tokio channel the response the sync world has to provide. This `await` operation is not blocking the tokio Runtime. This task will placed on pause and resumed as soon as a channel message is sent back to the async world

Thanks to async, the API can receive a really high number of request without getting stuck. The real limiting factor is inside of the sync world.

The worker pool is running inside of N+1 threads:
  * 1 thread is reserved to the code that receives all the requests to be evaluated. This is kinda like a phone operator from the old days
  * N threads: one per Wasm worker the user declared via the cli flag of PolicyServer

The "phone operator" from above works in this way:
  * It is always aware of which workers are busy processing a request and which ones are available
  * Once a message is received it finds the first available Wasm worker and dispatches to it the request to be evaluated

The "phone operator" is working in a non-blocking way. It never waits for the completion of a Wasm evaluation. It always reacts to events which can be of two types: an evaluation request is received, an evaluation result is ready.

Once the evaluation result is ready, the "phone operator" finds the channel associated with the API tokio task that generated the request and sends back the response of the tokio channel.

The API task is then resumed by the tokio runtime, the evaluation response is taken from the tokio channel and a HTTP response is sent back to the HTTP client that was waiting for it.

Having said that, the operation of "get the OCI manifest" involves two time expensive tasks:
  * A series of HTTP requests to fetch the manifest of an OCI image
  * Calculate the checksum of the manifest using the sha256 algorithm

These operation take some seconds, during which the Wasm worker is blocked waiting for an answer. This is a blocking operation, there's no other way to handle that because of how Wasm evaluation is done.
That means our "phone operator" can run temporarily run out of available Wasm workers if all of them are busy waiting for an OCI manifest to be calculated.

In our case, we had only 2 workers and 1 out of 4 policies causing these blocking "get me the OCI manifest" requests.

## The solution

There's no easy solution to this problem. Wasm evaluation cannot be done in an asynchronous way. The only workaround is to bite the bullet, but do that as seldom as possible.

The answer is to cache the result of the "get OCI manifest" operations, so that we pay this penalty a limited number of times.

This is what https://github.com/kubewarden/policy-evaluator/pull/67/commits/b6bcc6f67795ce255ec6eb36cae0c9cd83677896 does.

The cache we're using is time bound. Cached entries automatically expire after 60 seconds, and only successful results are cached. The latter point is a bit of a double sword: from one hand we don't want to a temporary hiccup with the network to cache an invalid results for 60 seconds, but from the other having a registry that is always down can cause a performance hit like the one we want to avoid.

## Testing the solution

I ran the load tests against a PolicyServer using this cache. I deliberately ran the code using a cache expiration of 10 seconds, instead of 60. I did that on purpose to simulate the impact of multiple cache misses.

The results I got have been super positive:

10 seconds cache

As you can see, we went back to almost normal numbers.

The numbers are even better with a 60 cache expiration:

60 seconds cache

## Overall

I think this is the way to go. It also makes sense IMO to cache this kind of value. I think 60 seconds is a reasonable amount of time.

**Comment by viccuad** [comment]
:heart:  Awesome work! and I learned a fair share. So IIUC, problem is that the operations of HTTP requests & calculating checksum  take too long, and never yield from the await. With a finite worker pool in policy-server (all sharing threads for async/await), all workers are busy and no work can proceed timely.

I think a cache is perfect and you provided an awesome justification! 

I wonder on consuming this in the policy-server side. Looking at https://docs.rs/tokio/1.14.0/tokio/index.html#cpu-bound-tasks-and-blocking-code, I wonder if we should use `tokio::task::spawn_blocking()` for `loop eval()` so they run in their own thread in their own thread pool. I must confess that I haven't tried it, but I feel that it would solve the worker blocking issue, as they liberate their thread to run a different await.
In the case that the digest is not in cache (first setup, or bunch of resources being created at once), could all the workers spend a lot of time and essentially block? (I suppose if one uses `spawn_blocking()` the cache needs to be thread-safe, and the used [cached](https://docs.rs/cached/latest/cached/) already is. And you have set `#[cached(sync_writes = true)]`).

Also, I wonder if the shutdown_channel in `loop_eval()` can be called once we have the digest, for those other async calls that are asking for the same digest. Maybe this to much busy work for little payoff.

I also thought about non-blocking versions of the http client such as reqwest (we talked about that at some point?). But even with having concurrent http requests, the workers will spend too much time and block. And this doesn't solve the checksum calculation. So we need the cache.

**Comment by flavio** [comment]
> So IIUC, problem is that the operations of HTTP requests & calculating checksum take too long, and never yield from the await. With a finite worker pool in policy-server (all sharing threads for async/await), all workers are busy and no work can proceed timely.

Not really. The HTTP/checksum operation are returning the control back to other tokio tasks. The problem is with the single wasm workers. they are not async (because wasmtime cannot work in this way), hence the workers are stuck in idle mode, waiting for the checkums to be returned to them.

> I wonder on consuming this in the policy-server side. Looking at https://docs.rs/tokio/1.14.0/tokio/index.html#cpu-bound-tasks-and-blocking-code, I wonder if we should use tokio::task::spawn_blocking() for loop eval() so they run in their own thread in their own thread pool. I must confess that I haven't tried it, but I feel that it would solve the worker blocking issue, as they liberate their thread to run a different await.

Based on what I wrote before, this would not help. What we would have to do is to allow the worker pool to grow dynamically based on the pressure we have on the queue. This is a future optimization we can do.

> In the case that the digest is not in cache (first setup, or bunch of resources being created at once), could all the workers spend a lot of time and essentially block? (I suppose if one uses spawn_blocking() the cache needs to be thread-safe, and the used cached already is. And you have set #[cached(sync_writes = true)]).

If two workers are asked to compute the same manifest digest and the cache is empty:
  * 1 worker will perform all the network/checksum requests
  * the other worker will wait for the value to be added to the cache, and then will resume operation

I think this approach is better than having two network requests done in parallel, 2 checksum computations done in parallel,...

> Also, I wonder if the shutdown_channel in loop_eval() can be called once we have the digest, for those other async calls that are asking for the same digest. Maybe this to much busy work for little payoff.

We have to let the loop eval run for as long as we are listening to the web API. The loop eval is not going to consume computation time. Quoting tokio's official docs:

```
> Tasks in Tokio are very lightweight. Under the hood, they require only a single allocation and 64 bytes of memory. Applications should feel free to spawn thousands, if not millions of tasks.
```

We need the loop to keep running because this is where all the "please compute this digest" messages that are sent by the Wasm workers are received and elaborated.

> I also thought about non-blocking versions of the http client such as reqwest (we talked about that at some point?). But even with having concurrent http requests, the workers will spend too much time and block. And this doesn't solve the checksum calculation. So we need the cache.

Exactly, you're spot on.

**Review by viccuad — APPROVED** [review]

File: src/callback_handler/mod.rs
Comment (flavio) [reply src/callback_handler/mod.rs:163]: I would leave everything cached for a little bit... I'm too afraid of performance issues. This can always be tuned later on

**Comment by flavio** [comment]
Merging. I just rebased a bit the history to make everything nicer

**Pull request merged**
